src/riak_kv_ts_newtype.erl

```diff
@@ -111,101 +113,133 @@ code_change(_OldVsn, State, _Extra) ->
 %%% Internal.
 %%%
 
-%%
 %% We rely on the claimant to not give us new DDLs after the bucket
 %% type is activated, at least until we have a system in place for
 %% managing DDL versioning
-do_new_type(BucketType) ->
-    maybe_compile_ddl(BucketType,
-                      retrieve_ddl_from_metadata(BucketType),
-                      riak_kv_compile_tab:get_ddl(BucketType),
-                      riak_ql_ddl_compiler:get_compiler_version(),
-                      riak_kv_compile_tab:get_compiled_ddl_version(BucketType)).
-
-maybe_compile_ddl(BucketType, NewDDL, NewDDL, NewVsn, NewVsn) ->
-    lager:info("Not compiling DDL for bucket type ~s because it is unchanged", [BucketType]),
-    %% Do nothing; we're seeing a CMD update but the DDL hasn't changed
-    ok;
-maybe_compile_ddl(BucketType, NewDDL, NewDDL, NewVsn, OldVsn) when is_record(NewDDL, ?DDL_RECORD_NAME),
-                                                                   is_integer(NewVsn), is_integer(OldVsn) ->
-    lager:info("Recompiling same DDL for bucket type ~s from version ~b to ~b",
-               [BucketType, OldVsn, NewVsn]),
-    actually_compile_ddl(BucketType, NewDDL);
-maybe_compile_ddl(BucketType, NewDDL, _OldDDL, NewVsn, OldVsn) when is_record(NewDDL, ?DDL_RECORD_NAME),
-                                                                    is_integer(OldVsn), is_integer(NewVsn) ->
-    lager:info("Compiling new DDL for bucket type ~s from version ~b to ~b",
-               [BucketType, OldVsn, NewVsn]),
-    actually_compile_ddl(BucketType, NewDDL);
-maybe_compile_ddl(BucketType, NewDDL, _OldDDL, NewVsn, notfound) when is_record(NewDDL, ?DDL_RECORD_NAME),
-                                                                      is_integer(NewVsn) ->
-    lager:info("Compiling new DDL for bucket type ~s with version ~b",
-               [BucketType, NewVsn]),
-    actually_compile_ddl(BucketType, NewDDL);
-maybe_compile_ddl(BucketType, NewDDL, _OldDDL, NewVsn, OldVsn) ->
-    lager:error("Unknown DDL version type ~p for bucket type ~s "
-                "(expecting ~p with old version ~p, new version ~p)",
-                [NewDDL, BucketType, ?DDL_RECORD_NAME, OldVsn, NewVsn]),
-    %% We don't know what to do with this new DDL, so stop
+do_new_type(Table) ->
+    case retrieve_ddl_from_metadata(Table) of
+        undefined ->
+            %% this bucket type name does not exist in the metadata!
+            log_missing_ddl_metadata(Table);
+        DDL ->
+            CurrentVersion = riak_ql_ddl:current_version(),
+            case riak_kv_compile_tab:get_ddl(Table, CurrentVersion) of
```

Comment:
What is in the metadata might not be a DDL record type we can deal with, so prefer the one in the dets table if it exists which means the node has been downgraded. If it is not in dets then it is a truly new table.  It is safe to compile a new beam and load it as many teams as is necessary.